trin-core/src/portalnet/overlay.rs

```diff
@@ -192,12 +201,160 @@ where
+                    content_keys_values.push((key, content_value))
+                }
+                Err(err) => {
+                    return Err(anyhow!(
+                        "Unexpected error while decoding overlay content key: {err}"
+                    ));
+                }
+            }
+        }
+        // Propagate gossip accepted content
+        self.propagate_gossip(content_keys_values);
+        Ok(())
+    }
+
+    /// Propagate gossip accepted content via OFFER/ACCEPT:
+    fn propagate_gossip(&self, content: Vec<(TContentKey, ByteList)>) {
+        // Get all nodes from overlay routing table
+        let kbuckets = self.kbuckets.read();
+        let all_nodes: Vec<&kbucket::Node<NodeId, Node>> = kbuckets
+            .buckets_iter()
+            .map(|kbucket| {
+                kbucket
+                    .iter()
+                    .collect::<Vec<&kbucket::Node<NodeId, Node>>>()
+            })
+            .flatten()
+            .collect();
+
+        // HashMap to temporarily store all interested ENRs and the content.
+        // Key is base64 string of node's ENR.
+        let mut enrs_and_content: HashMap<String, Vec<RawContentKey>> = HashMap::new();
+
+        // Filter all nodes from overlay routing table where XOR_distance(content_id, nodeId) < node radius
+        for content_key_value in content {
+            let interested_enrs: Vec<Enr> = all_nodes
+                .clone()
+                .into_iter()
+                .filter(|node| {
+                    XorMetric::distance(
+                        &content_key_value.0.content_id(),
+                        &node.key.preimage().raw(),
+                    ) < node.value.data_radius()
+                })
+                .map(|node| node.value.enr())
+                .collect();
+
+            // Continue if no nodes are interested in the content
+            if interested_enrs.is_empty() {
+                debug!("No nodes interested in neighborhood gossip: content_key={} num_nodes_checked={}",
+                    hex_encode(content_key_value.0.into()), all_nodes.len());
+                continue;
+            }
+
+            // Get log2 random ENRs to gossip. This should be safe unwrap as we continue the loop
+            // above if no nodes are interested in the content.
+            let random_enrs = log2_random_enrs(interested_enrs).unwrap();
+
+            // Temporarily store all randomly selected nodes with the content of interest.
+            // We want this so we can offer all the content to interested node in one request.
+            for enr in random_enrs {
+                enrs_and_content
+                    .entry(enr.to_base64())
+                    .or_default()
+                    .push(content_key_value.clone().0.into());
+            }
+        }
+
+        // Create and send OFFER overlay request to the interested nodes
+        for (enr_string, interested_content) in enrs_and_content.drain() {
```

Comment:
I think we can consume the map here with `into_iter`. I don't see why we need to preserve the allocated memory with `drain`.